diff --git a/src/Observers/EeMediaGalleryUpdateObserver.php b/src/Observers/EeMediaGalleryUpdateObserver.php
index eaf339e..1489ca6 100644
--- a/src/Observers/EeMediaGalleryUpdateObserver.php
+++ b/src/Observers/EeMediaGalleryUpdateObserver.php
@@ -34,6 +34,29 @@ use TechDivision\Import\Product\Media\Ee\Utils\MemberNames;
 class EeMediaGalleryUpdateObserver extends EeMediaGalleryObserver
 {
 
+    /**
+     * Initialize the product media gallery with the passed attributes and returns an instance.
+     *
+     * @param array $attr The product media gallery attributes
+     *
+     * @return array The initialized product media gallery
+     */
+    protected function initializeProductMediaGallery(array $attr)
+    {
+
+        // load the value and the attribute ID
+        $value = $attr[MemberNames::VALUE];
+        $attributeId = $attr[MemberNames::ATTRIBUTE_ID];
+
+        // query whether the product media gallery entity already exists or not
+        if ($entity = $this->loadProductMediaGallery($attributeId, $value)) {
+            return $this->mergeEntity($entity, $attr);
+        }
+
+        // simply return the attributes
+        return $attr;
+    }
+
     /**
      * Initialize the product media gallery value to entity with the passed attributes and returns an instance.
      *
@@ -57,6 +80,19 @@ class EeMediaGalleryUpdateObserver extends EeMediaGalleryObserver
         return $attr;
     }
 
+    /**
+     * Load's the product media gallery with the passed attribute ID + value.
+     *
+     * @param integer $attributeId The attribute ID of the product media gallery to load
+     * @param string  $value       The value of the product media gallery to load
+     *
+     * @return array The product media gallery
+     */
+    protected function loadProductMediaGallery($attributeId, $value)
+    {
+        return $this->getProductMediaProcessor()->loadProductMediaGallery($attributeId, $value);
+    }
+
     /**
      * Load's the product media gallery with the passed value/entity ID.
      *
diff --git a/src/Observers/EeMediaGalleryValueObserver.php b/src/Observers/EeMediaGalleryValueObserver.php
index 7e7850f..cc05a0e 100644
--- a/src/Observers/EeMediaGalleryValueObserver.php
+++ b/src/Observers/EeMediaGalleryValueObserver.php
@@ -58,6 +58,9 @@ class EeMediaGalleryValueObserver extends MediaGalleryValueObserver
         // load the image label
         $imageLabel = $this->getValue(ColumnKeys::IMAGE_LABEL);
 
+        // load the flag that decides whether or not an image should be hidden on product page
+        $hideFromProductPage = $this->getValue(ColumnKeys::HIDE_FROM_PRODUCT_PAGE);
+
         // prepare the media gallery value
         return $this->initializeEntity(
             array(
@@ -66,7 +69,7 @@ class EeMediaGalleryValueObserver extends MediaGalleryValueObserver
                 MemberNames::ROW_ID      => $parentId,
                 MemberNames::LABEL       => $imageLabel,
                 MemberNames::POSITION    => $position,
-                MemberNames::DISABLED    => 0
+                MemberNames::DISABLED    => $hideFromProductPage
             )
         );
     }
